File: 1. 단계별로_풀어보기/20~29/25. 이분 탐색/6. 1300.py
import sys

input = sys.stdin.readline
output = sys.stdout.write

N = int(input())
K = int(input())

def findLessCnt(val):
  global N
  result = 0
  for i in range(1, N + 1):
    if (val - 1) >= N * i:
      result += N
    elif val - 1 >= i:
      result += (val - 1) // i
    else:
      break
  return result

def binarySearch(start, end):
  global K

  if start == end:
    return start
  if start + 1 == end:
    if findLessCnt(end) <= K - 1:
      return end
    return start
  
  mid = (start + end) // 2
  
  lessThenMid = findLessCnt(mid)
  
  if lessThenMid > K - 1:
    return binarySearch(start, mid - 1)
  else:
    return binarySearch(mid, end)
  
output(f"{binarySearch(0, min(10 ** 9, N * N))}")

File: 1. 단계별로_풀어보기/20~29/25. 이분 탐색/6. 1300.js
// 입력
const fs = require('fs')
const inputArr = fs.readFileSync(0).toString().trim().split('\n')

// 입력 파싱
const N = BigInt(inputArr[0])
const K = BigInt(inputArr[1])

// 결과 연산
let result = binarySearch(0n, 10000000000n > N * N ? N * N : 1000000000n)

process.stdout.write(`${result}`)

function binarySearch(minVal, maxVal) {
  if (minVal === maxVal) {
    return minVal
  }

  if (minVal + 1n === maxVal) {
    if (findLessCnt(maxVal) < K) {
      return maxVal
    }
    return minVal
  }

  const midVal = (minVal + maxVal) / 2n

  if (findLessCnt(midVal) > K - 1n) {
    return binarySearch(minVal, midVal - 1n)
  } // ::
  else {
    return binarySearch(midVal, maxVal)
  }
}

function findLessCnt(val) {
  let result = 0n
  for (let i = 1n; i <= N; i++) {
    if (val > N * i) {
      result += N
    } // ::
    else if (val > i) {
      result += (val - 1n) / i
    } // ::
    else {
      break
    }
  }

  return result
}
